src/slurm_plugin/cluster_event_publisher.py

```diff
@@ -0,0 +1,371 @@
+    def publish_event(self):
+        return self._publish_event
+
+    @staticmethod
+    def timestamp():
+        return datetime.now(timezone.utc).isoformat(timespec="milliseconds")
+
+    @log_exception(logger, "publish_unhealthy_static_node_events", catch_exception=Exception, raise_on_error=False)
+    def publish_unhealthy_static_node_events(
+        self,
+        unhealthy_static_nodes: List[SlurmNode],
+        nodes_in_replacement: Iterable[str],
+        failed_nodes: Dict[str, List[str]],
+    ):
+        timestamp = ClusterEventPublisher.timestamp()
+
+        nodes_in_replacement = list(nodes_in_replacement)
+
+        self.publish_event(
+            "WARNING" if failed_nodes else "DEBUG",
+            "Number of static nodes that failed replacement after node maintenance",
+            event_type="node-launch-failure-count",
+            timestamp=timestamp,
+            event_supplier=self._get_launch_failure_details(failed_nodes),
+        )
+
+        self.publish_event(
+            "DEBUG",
+            "After node maintenance, node failed replacement",
+            event_type="node-launch-failure",
+            timestamp=timestamp,
+            event_supplier=self._failed_node_supplier(unhealthy_static_nodes, failed_nodes),
+        )
+
+        self.publish_event(
+            "DEGUG",
+            "Number of static nodes failing scheduler health check",
+            event_type="static-node-health-check-failure-count",
+            timestamp=timestamp,
+            event_supplier=self._unhealthy_node_count_supplier(unhealthy_static_nodes),
+        )
+
+        self.publish_event(
+            "DEBUG",
+            "Static node failing scheduler health check",
+            event_type="static-node-health-check-failure",
+            timestamp=timestamp,
+            event_supplier=self._unhealthy_node_supplier(unhealthy_static_nodes),
+        )
+
+        self.publish_event(
+            "DEBUG",
+            "Number of instances terminated due to backing unhealthy static nodes",
+            event_type="static-node-instance-terminate-count",
+            timestamp=timestamp,
+            event_supplier=self._terminated_instances_supplier(
+                (node for node in unhealthy_static_nodes if node.instance)
+            ),
+        )
+
+        self.publish_event(
+            "DEBUG",
+            "After node maintenance, nodes currently in replacement",
+            event_type="static-nodes-in-replacement-count",
+            timestamp=timestamp,
+            event_supplier=self._node_count_in_replacement_supplier(nodes_in_replacement),
+        )
+
+        self.publish_event(
+            "DEBUG",
+            "After node maintenance, node currently in replacement",
+            event_type="static-node-in-replacement",
+            timestamp=timestamp,
+            event_supplier=self._node_in_replacement_supplier(
+                (node for node in unhealthy_static_nodes if node.name in nodes_in_replacement)
```

Comment:
Is this event try to represent all static nodes that are currently in replacement after relaunching unhealthy static ndoes? Why these are nodes in unhealthy_static_nodes? Should it be just`node.name in nodes_in_replacement`?

Reply:
the _nodes_in_replacement_supplier should be logging out the full node description, so it needs the SlurmNode rather than just the node.name.